### PR TITLE
Cancun support

### PR DESCRIPTION
**Pull request by RomarQ**
- [x] Adds 2 new opcodes (`TLOAD` and `TSTORE`) for manipulating state that behaves almost identically to storage but is discarded after every transaction.

    - These new instructions were introduced by [EIP-1153](https://eips.ethereum.org/EIPS/eip-1153) and included in Cancun fork.

- [x] Adds a new opcode  (`MCOPY`) for copying memory. Implementing this change will be important in ensuring that the same solidity/vyper contract compiles down to the same bytecode.´
   - This new instruction was introduced by [EIP-5656](https://eips.ethereum.org/EIPS/eip-5656) and included in Cancun fork.

- [x] Changes the behaviour of the self destruct opcode.
   - [EIP-6780: SELFDESTRUCT only in same transaction](https://eips.ethereum.org/EIPS/eip-6780)

### CODE REVIEW

File: frame/evm/src/lib.rs [hunk 1]
```diff
@@ -821,32 +821,14 @@ impl<T: Config> Pallet<T> {
...
-
-					let account_id = T::AddressMapping::into_account_id(*address);
-					let _ = frame_system::Pallet::<T>::dec_sufficients(&account_id);
-				}
-				KillStorageResult::SomeRemaining(_) => (),
-			}
-		}
+		#[allow(deprecated)]
+		let _ = <AccountStorages<T>>::remove_prefix(address, None);
 	}
```
Comment (ahmadkaouk) [inline frame/evm/src/lib.rs:832]: After EIP-6780, contract deletion is only permitted within the same transaction that created it. This means it is now safe, from a PoV perspective, to remove the storage entries once the contract is deleted. There is no need anymore to mark it as suicided and remove the entries afterward.

File: frame/evm/src/lib.rs [hunk 2]
```diff
@@ -821,32 +821,14 @@ impl<T: Config> Pallet<T> {
 	/// Remove an account.
 	pub fn remove_account(address: &H160) {
 		if <AccountCodes<T>>::contains_key(address) {
-			// Remember to call `dec_sufficients` when clearing Suicided.
-			<Suicided<T>>::insert(address, ());
```
Comment (RomarQ) [inline frame/evm/src/lib.rs:825]: Not sure about this change. How will this affect the heuristics of previous "corrupted" suicided contracts?

File: frame/evm/src/lib.rs [hunk 3]
```diff
@@ -821,32 +821,14 @@ impl<T: Config> Pallet<T> {
...
-					<Suicided<T>>::remove(address);
-
-					let account_id = T::AddressMapping::into_account_id(*address);
-					let _ = frame_system::Pallet::<T>::dec_sufficients(&account_id);
-				}
-				KillStorageResult::SomeRemaining(_) => (),
-			}
-		}
+		#[allow(deprecated)]
+		let _ = <AccountStorages<T>>::remove_prefix(address, None);
```
Comment (RomarQ) [inline frame/evm/src/lib.rs:831]: Lets avoid using this deprecated method

```suggestion
		let _ = <AccountStorages<T>>::clear_prefix(address, u32::max_value(), None);
```

File: frame/evm/src/lib.rs (hunk 1 above)
Comment (RomarQ) [reply frame/evm/src/lib.rs:832]: I understand this. Would be nice to have a test for this, were we try to self-destruct a big contract that already existed.

File: frame/evm/src/lib.rs (hunk 2 above)
Comment (RomarQ) [reply frame/evm/src/lib.rs:825]: Ignore my previous comment, but we can probably get rid of all the logic related to `clear_suicided_contract`.

Comment (ahmadkaouk) [reply frame/evm/src/lib.rs:825]: Yes, with the new `selfdestruct` changes, we no longer need the suicide contract logic. However, we cannot remove all related logic before ensuring that current suicided contracts and their storage are properly removed using the storage-cleaner precompile, and that the corrupted suicided contract is also cleaned through the lazy migration extrinsic.

**Review by librelois — APPROVED** [review]

**Pull request merged**